packages/fluentui/react-component-ref/src/Ref.tsx

```diff
@@ -1,16 +1,159 @@
+// react/packages/react-reconciler/src/ReactFiber.js
+// ========================================================
+
+type Fiber = {
+  // Tag identifying the type of fiber.
+  tag: string;
+  // The resolved function/class/ associated with this fiber.
+  type: any;
 };
+
+/**
+ * Detects if a passed element is a Fiber object instead of an element. Is needed as `ReactDOM.findDOMNode()` returns
+ * a Fiber in `react-test-renderer` that can cause issues with tests. Is used only in non-production env.
+ *
+ * @see https://github.com/facebook/react/issues/7371#issuecomment-317396864
+ * @see https://github.com/Semantic-Org/Semantic-UI-React/issues/4061#issuecomment-694895617
+ */
+function isFiberRef(node: Element | Fiber | Text | null): boolean {
+  if (node === null) {
+    return false;
+  }
+
+  if (node instanceof Element || node instanceof Text) {
+    return false;
+  }
+
+  return !!(node.type && node.tag);
+}
+
+export class Ref extends React.Component<RefProps, RefState> {
+  prevNode?: Node | null;
+  currentNode?: Node | null;
+
+  state = { kind: null };
+
+  static getDerivedStateFromProps(props: RefProps): RefState {
+    const child = React.Children.only(props.children);
+
+    if (child.type === Ref) {
+      return { kind: 'self' };
+    }
+
+    if (ReactIs.isForwardRef(child)) {
+      return { kind: 'forward' };
+    }
+
+    return { kind: 'find' };
+  }
+
+  handleForwardRefOverride = (node: HTMLElement) => {
+    const { children, innerRef } = this.props;
+
+    handleRef((children as React.ReactElement & { ref: React.Ref<any> }).ref, node);
+    handleRef(innerRef, node);
+
+    this.currentNode = node;
+  };
+
+  handleSelfOverride = (node: HTMLElement) => {
+    const { children, innerRef } = this.props;
+
+    handleRef((children as React.ReactElement<RefProps>).props.innerRef, node);
+    handleRef(innerRef, node);
+  };
+
+  componentDidMount() {
+    if (this.state.kind === 'find') {
+      let currentNode = ReactDOM.findDOMNode(this);
+
+      if (process.env.NODE_ENV !== 'production') {
+        if (isFiberRef(currentNode)) {
+          currentNode = null;
+        }
+      }
+
+      this.prevNode = currentNode;
+      handleRef(this.props.innerRef, currentNode);
+    }
+  }
+
+  componentDidUpdate(prevProps: RefProps) {
+    if (this.state.kind === 'forward') {
+      if (prevProps.innerRef !== this.props.innerRef) {
+        handleRef(this.props.innerRef, this.currentNode);
+      }
+    } else if (this.state.kind === 'find') {
+      let currentNode = ReactDOM.findDOMNode(this);
+
+      if (process.env.NODE_ENV !== 'production') {
+        if (isFiberRef(currentNode)) {
+          currentNode = null;
+        }
+      }
+
+      if (this.prevNode !== currentNode) {
+        this.prevNode = currentNode;
+        handleRef(this.props.innerRef, currentNode);
+      }
+
+      if (prevProps.innerRef !== this.props.innerRef) {
+        handleRef(this.props.innerRef, currentNode);
```

Comment:
If both `innerRef` and `currentNode` change, the `handleRef()` is called twice.

Reply:
Good catch, I will check it and cover it with tests